WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/Time Range/AnalyticsHubTimeRangeSelection.swift

```diff
@@ -0,0 +1,138 @@
+    var currentTimeRange: AnalyticsHubTimeRange? { get }
+    var previousTimeRange: AnalyticsHubTimeRange? { get }
+    var currentRangeDescription: String? { get }
+    var previousRangeDescription: String? { get }
+
+    init(referenceDate: Date, currentCalendar: Calendar)
+}
+
+/// Main source of time ranges of the Analytics Hub, responsible for providing the current and previous dates
+/// for a given Date and range Type alongside their UI descriptions
+///
+public class AnalyticsHubTimeRangeSelection {
+    private let selectionDelegate: AnalyticsHubTimeRangeSelectionDelegate
+    let rangeSelectionDescription: String
+
+    //TODO: abandon usage of the ISO 8601 Calendar and build one based on the Site calendar configuration
+    init(selectionType: SelectionType,
+         currentDate: Date = Date(),
+         currentCalendar: Calendar = Calendar(identifier: .iso8601)) {
+        self.rangeSelectionDescription = selectionType.description
+
+        switch selectionType {
+        case .today:
+            self.selectionDelegate = AnalyticsHubDayRangeSelection(referenceDate: currentDate, currentCalendar: currentCalendar)
+        case .weekToDate:
+            self.selectionDelegate = AnalyticsHubWeekRangeSelection(referenceDate: currentDate, currentCalendar: currentCalendar)
+        case .monthToDate:
+            self.selectionDelegate = AnalyticsHubMonthRangeSelection(referenceDate: currentDate, currentCalendar: currentCalendar)
+        case .yearToDate:
+            self.selectionDelegate = AnalyticsHubYearRangeSelection(referenceDate: currentDate, currentCalendar: currentCalendar)
+        }
+    }
+
+    /// Unwrap the generated selected `AnalyticsHubTimeRange` based on the `selectedTimeRange`
+    /// provided during initialization.
+    /// - throws an `.selectedRangeGenerationFailed` error if the unwrap fails.
+    func unwrapCurrentTimeRange() throws -> AnalyticsHubTimeRange {
+        guard let currentTimeRange = selectionDelegate.currentTimeRange else {
+            throw TimeRangeGeneratorError.selectedRangeGenerationFailed
+        }
+        return currentTimeRange
+    }
+
+    /// Unwrap the generated previous `AnalyticsHubTimeRange`relative to the selected one
+    /// based on the `selectedTimeRange` provided during initialization.
+    /// - throws a `.previousRangeGenerationFailed` error if the unwrap fails.
+    func unwrapPreviousTimeRange() throws -> AnalyticsHubTimeRange {
+        guard let previousTimeRange = selectionDelegate.previousTimeRange else {
+            throw TimeRangeGeneratorError.previousRangeGenerationFailed
+        }
+        return previousTimeRange
+    }
+
+    /// Generates a date description of the previous time range set internally.
+    /// - Returns the Time range in a UI friendly format. If the previous time range is not available,
+    /// then returns an presentable error message.
+    func generateCurrentRangeDescription() -> String {
```

Comment:
Since it just provides convenient access for the variable - can we make it computed `var currentRangeDescription` instead of a `func`?
From `func` API I expect some processing inside or additional side effects.

Reply:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8267/commits/5f4ad85a17f30cee5f93f6cb580dab4b1844e064